### PR TITLE
optimizations

### PR DESCRIPTION
**Pull request by Ilaiwi**
Improvements to eliminate extra components renders. Please refer to #303 for extra information.

An additional improvement would be removing `deepObjectCompare` from `Items` and `Item` and use [immer](https://github.com/mweststrate/immer) when calculating `dimensionItems`, `groupHeights` and `groupTops`

### CODE REVIEW

**Comment by Ilaiwi** [comment]
@mcMickJuice I stumbled on a new library that claims to be 3x faster and smaller than `lodash.isEqual` which is [react-fast-compare](https://github.com/FormidableLabs/react-fast-compare). Maybe replacing lodash with this one?

File: demo/index.js [hunk 1]
```diff
@@ -5,6 +5,11 @@ import App from './app'
 
 import './index.html'
 
+if (process.env.NODE_ENV !== 'production') {
+  const {whyDidYouUpdate} = require('why-did-you-update');
+  whyDidYouUpdate(React);
```
Comment (mcMickJuice) [inline demo/index.js]: do you mind removing this block?  This should only be used to profile performance as a means to an end.  I don't want to add a dependency that profiles performance (i.e. it should only be used when performing dev work)

File: package.json [hunk 2]
```diff
@@ -125,6 +125,7 @@
     "sass-loader": "~4.0.2",
     "style-loader": "~0.13.0",
     "webpack": "~1.13.3",
-    "webpack-dev-server": "~1.16.2"
+    "webpack-dev-server": "~1.16.2",
+    "why-did-you-update": "^0.1.1"
```
Comment (mcMickJuice) [inline package.json]: please remove why-did-you-update

File: src/lib/Timeline.js [hunk 3]
```diff
@@ -947,6 +952,32 @@ export default class ReactCalendarTimeline extends Component {
     return label ? <InfoLabel label={label} /> : ''
   }
 
+  renderLeftSidebar = () => {
```
Comment (mcMickJuice) [inline src/lib/Timeline.js]: Why move these into separate functions? Is this related to optimizations?

File: src/lib/lines/TodayLine.js [hunk 4]
```diff
@@ -12,10 +12,26 @@ export default class TodayLine extends Component {
   }
 
   static defaultProps = {}
+  
+  state = {
+    currentTime: new Date(),
+  }
+
+  componentDidMount(){
+    this.interval = setInterval(()=> {
```
Comment (mcMickJuice) [inline src/lib/lines/TodayLine.js]: Are these changes specific to optimizations?  If not, can you please remove these changes.  I'd like to keep PRs focused in their intent.  

I see where you're going here and I agree with these changes but they should be in a separate PR.

**Comment by mcMickJuice** [comment]
Is the reason you've added deepCompare and arraysEqual because the objects `dimensionItems`, `groupHeights` and `groupTops` are new references every time we render?  If this is the case, we should look at the methods that create these values and ensure that THOSE don't return new references.

deepCompares are expensive and I really don't want to do them to be called in a critical path such as rendering.  We should instead look to memoize the functions that return these values and only return a new reference if an input has changed.

When you profiled this component, in what situations were we unnecessarily rendering?  Was it related only to dragging items?  Or was it all interactions?

File: src/lib/Timeline.js (hunk 3 above)
Comment (Ilaiwi) [reply src/lib/Timeline.js]: Yes, it is for optimization reasons since everytime you call the `head` method a new object for `leftSidebar` and `rightSidebar` will get generated and the `Header` is a pure component which will result in extra renders

File: src/lib/lines/TodayLine.js (hunk 4 above)
Comment (Ilaiwi) [reply src/lib/lines/TodayLine.js]: Well, this component as well had a lot of extra renders where it got rendered every time out timeline rendered which is not optimal. When I changed it to a pure component this resulted in it being never updated as expected since no state nor props where updated although time changed and it should've moved so it was more of solving a bug. Does this change worth being in a separate pull request?

**Comment by Ilaiwi** [comment]
I agree `deepEqual` and `arrayEqual` are expensive operations but not as expensive as rendering so using them resulted in a much faster and snappier calendar. 
I will look into your suggestion regarding `dimensionItems`, `groupHeights` and `groupTops`. I looked into it the first time and honestly, I don't remember why I didn't do them as mentioned.
Yes, extra renders were caused by all interactions. I really encourage you to install `why-did-you-update` and see for yourself and also try the component after optimization.

**Comment by mcMickJuice** [comment]
Hi @Ilaiwi I really appreciate the effort you put in here however I'm going to close this PR.  I'd like to take a different approach to how this library manages internal state which will change how each component receives props.

I'm definitely going to draw inspiration from these proposals though and make sure to use the why-did-you-update library!

**Comment by Ilaiwi** [comment]
@mcMickJuice you are welcome please let me know if i can help.
